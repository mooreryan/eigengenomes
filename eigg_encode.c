#include "eigg_encode.h"

#include <assert.h>
#include <stdlib.h>
#include <ctype.h>

double* eigg_encode_ary_init()
{
  double* ary = calloc('T' + 1, sizeof(double));
  assert(ary != NULL);

  ary['A'] =  1;
  ary['C'] =  0.5;
  ary['T'] = -1;
  ary['G'] = -0.5;
  ary['N'] =  0;

  return ary;
}

void eigg_encode_ary_destroy(double* ary)
{
  free(ary);
}

double* eigg_encode_kmer(double* encoded_kmer,
                         double* encode_nt,
                         char* kmer,
                         unsigned long kmer_len)
{
  unsigned long i = 0;

  for (i = 0; i < kmer_len; ++i) {
    /* TODO assert that kmer[i] is a valid AA */
    encoded_kmer[i] = encode_nt[toupper(kmer[i])];
  }

  return encoded_kmer;
}
