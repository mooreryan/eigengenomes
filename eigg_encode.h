#ifndef EIGG_ENCODE_H
#define EIGG_ENCODE_H

double* eigg_encode_ary_init();

void eigg_encode_ary_destroy(double* ary);

double* eigg_encode_kmer(double* encoded_kmer,
                         double* encode_nt,
                         char* kmer,
                         unsigned long kmer_len);

#endif
