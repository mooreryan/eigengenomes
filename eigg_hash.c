#include "eigg_hash.h"

double
eigg_hash_encoded_kmer(double** hyperplanes,
                       int      num_hyperplanes,
                       double*  encoded_kmer,
                       int      kmer_len,
                       double*  pow2)
{
  int     hyperplane_i = 0;
  int     kmer_i = 0;
  double* this_hyperplane;
  double hash_val = 0.0;

  for (hyperplane_i = 0; hyperplane_i < num_hyperplanes; ++hyperplane_i) {
    this_hyperplane = hyperplanes[hyperplane_i];

    double sum = 0.0;

    for (kmer_i = 0; kmer_i < kmer_len - 1; ++kmer_i) {
      sum = sum + (encoded_kmer[kmer_i] * this_hyperplane[kmer_i]);
    }

    /* subtract the last...plane through origin */
    sum = sum - (encoded_kmer[kmer_i] * this_hyperplane[kmer_i]);

    if (sum > 0) {
      hash_val += pow2[hyperplane_i];
    }

    /* if (sum > 0) { */
    /*   hashed_kmer[hyperplane_i] = '1'; */
    /* } else { */
    /*   hashed_kmer[hyperplane_i] = '0'; */
    /* } */
  }

  return hash_val;
}
