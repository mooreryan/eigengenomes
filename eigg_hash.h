#ifndef EIGG_HASH_H
#define EIGG_HASH_H

double
eigg_hash_encoded_kmer(double** hyperplanes,
                       int      num_hyperplanes,
                       double*  encoded_kmer,
                       int      kmer_len,
                       double*  pow2);

#endif
