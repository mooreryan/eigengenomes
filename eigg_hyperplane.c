#include "eigg_hyperplane.h"

#include <stdlib.h>

static double
rand_real()
{
  /* TODO switch to random(4) for better random-ness? */
  return (double) rand() / (double) RAND_MAX;
}

/* random hyperplane through the origin */
static double*
hyperplane_init(int num_coef)
{
  double* hyperplane = malloc(num_coef * sizeof(double));

  int i = 0;
  double coef = 0.0;

  for (i = 0; i < num_coef; ++i) {
    coef = rand_real();

    /* if (rand_real() > 0.5) { */
    /*   coef = 1 / coef; */
    /* } */

    if (rand_real() > 0.5) {
      coef = -coef;
    }

    hyperplane[i] = coef;
  }

  return hyperplane;
}

static void
hyperplane_destroy(double* hyperplane)
{
  free(hyperplane);
}

/* num_coef is normally the kmer len */
double**
eigg_hyperplane_ary_init(int num_hyperplanes,
                         int num_coef)
{
  int i = 0;

  double** hyperplane_ary = malloc(num_hyperplanes * sizeof(double*));

  for (i = 0; i < num_hyperplanes; ++i) {
    hyperplane_ary[i] = hyperplane_init(num_coef);
  }

  return hyperplane_ary;
}

void
eigg_hyperplane_ary_destroy(double** hyperplane_ary,
                            int num_hyperplanes)
{
  int i = 0;

  for (i = 0; i < num_hyperplanes; ++i) {
    hyperplane_destroy(hyperplane_ary[i]);
  }

  free(hyperplane_ary);
}
