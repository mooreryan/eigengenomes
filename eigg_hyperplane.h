#ifndef EIGG_HYPERPLANE_H
#define EIGG_HYPERPLANE_H

double** eigg_hyperplane_ary_init(int num_hyperplanes, int num_coef);
void eigg_hyperplane_ary_destroy(double** hyperplane_ary, int num_hyperplanes);

#endif
