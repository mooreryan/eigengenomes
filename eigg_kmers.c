#include "eigg_kmers.h"

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>

static long get_num_kmers(long seq_len, unsigned long kmer_len)
{
  return seq_len - kmer_len + 1;
}

static char** get_kmers(char* seq, unsigned long kmer_len, long num_kmers)
{
  long i = 0;
  unsigned long  j = 0;

  char** kmers = malloc(num_kmers * sizeof(char*));

  for (i = 0; i < num_kmers; ++i) {
    char* kmer = malloc((kmer_len + 1) * sizeof(char));

    for (j = 0; j < kmer_len; ++j) {
      kmer[j] = seq[i + j];
    }
    kmer[j] = '\0';

    kmers[i] = kmer;
  }

  return kmers;
}

struct Kmers* eigg_kmers_new(char* seq, long seq_len, unsigned long kmer_len)
{
  struct Kmers* kmers = malloc(sizeof(struct Kmers));
  assert(kmers != NULL);

  long num_kmers = get_num_kmers(seq_len, kmer_len);

  kmers->kmers     = get_kmers(seq, kmer_len, num_kmers);
  kmers->num_kmers = num_kmers;
  kmers->kmer_len  = kmer_len;

  return kmers;
}

void eigg_kmers_destroy(struct Kmers* kmers)
{
  long i = 0;

  for (i = 0; i < kmers->num_kmers; ++i) {
    free(kmers->kmers[i]);
  }

  free(kmers->kmers);
  free(kmers);
}

void eigg_kmers_print(struct Kmers* kmers)
{
  long i = 0;

  for (i = 0; i < kmers->num_kmers; ++i) {
    printf("#%ld -- %s\n", i, kmers->kmers[i]);
  }
}
