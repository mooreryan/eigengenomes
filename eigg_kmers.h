#ifndef EIGG_KMERS_H
#define EIGG_KMERS_H

struct Kmers {
  char** kmers;
  long num_kmers;
  unsigned long kmer_len;
};

struct Kmers* eigg_kmers_new(char* seq, long seq_len, unsigned long kmer_len);

void eigg_kmers_destroy(struct Kmers* kmers);

void eigg_kmers_print(struct Kmers* kmers);

#endif
