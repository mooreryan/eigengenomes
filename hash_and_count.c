#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <zlib.h>

#include "kseq.h"
#include "eigg_encode.h"
#include "eigg_hash.h"
#include "eigg_hyperplane.h"
#include "eigg_kmers.h"
#include "eigg_print.h"
#include "eigg_version.h"

#include "vendor/tommyarray.h"
#include "vendor/tommyhashlin.h"

KSEQ_INIT(gzFile, gzread)

struct hash_bucket_count_t {
  unsigned long hash_bucket;
  unsigned long count;

  tommy_node node;
};

int hash_bucket_count_compare(const void* arg, const void* hash_bucket_count)
{
  return *(const unsigned long*)arg !=
    ((const struct hash_bucket_count_t*)hash_bucket_count)->hash_bucket;
}

/* Input a kseq, and print it out fastA/Q style */
void print_record(kseq_t* seq)
{
  if (seq->qual.l) {
    putchar('@');
  } else {
    putchar('>');
  }

  printf("%s", seq->name.s);

  if (seq->comment.l) {
    printf(" %s\n", seq->comment.s);
  } else {
    putchar('\n');
  }

  printf("%s\n", seq->seq.s);

  if (seq->qual.l) { printf("+\n%s\n", seq->qual.s); }
}

int main(int argc, char* argv[])
{


  long l;
  unsigned long i = 0;
  unsigned long kmer_len = 0;
  unsigned long num_seen_hash_buckets = 0;
  unsigned long num_hyperplanes = 0;
  unsigned long num_seqs = 0;
  long seed = 0;

  kseq_t* seq;
  gzFile  fp;

  double** hyperplanes;
  double* encode_nt;
  double* encoded_kmer;

  unsigned long hashed_kmer;

  char* s1_name;
  char* s1_seq;
  char* s1_qual;
  unsigned long s1_seq_len = 0;

  tommy_array*   hash_bucket_names = malloc(sizeof(tommy_array));
  assert(hash_bucket_names != NULL);
  tommy_hashlin* hash_bucket_counts = malloc(sizeof(tommy_hashlin));
  assert(hash_bucket_counts != NULL);

  /* to hold the queries of the hash table */
  struct hash_bucket_count_t* hash_bucket_count;

  if (argc != 5) {
    fprintf(stderr,
            "VERSION: %s\nUsage: %s "
            "<1: seed> <2: kmer size> "
            "<3: number of hyperplanes> <4: seqs.fastq> "
            "> seqs.hash_counts\n",
            VERSION,
            argv[0]);

    return 1;
  }

  fp = gzopen(argv[4], "r");
  if (!fp) {
    fprintf(stderr, "ERROR -- could not open %s\n", argv[4]);

    return 2;
  }

  seed = strtol(argv[1], NULL, 10);
  srand(seed);

  /* TODO toss out the first few random values */

  kmer_len = strtol(argv[2], NULL, 10);

  encoded_kmer = malloc(kmer_len * sizeof(double));
  assert(encoded_kmer != NULL);

  /* Total bins = 2^num_hyperplanes */
  num_hyperplanes = strtol(argv[3], NULL, 10);


  unsigned long num_hash_buckets =
    (unsigned long)pow(2, num_hyperplanes);

  /* unsigned long* hashed_kmer_counts = */
  /*   calloc(num_hash_buckets, sizeof(unsigned long)); */
  /* assert(hashed_kmer_counts != NULL); */

  double* pow2 = malloc(num_hyperplanes * sizeof(double));
  assert(pow2 != NULL);
  for (int i = 0; i < num_hyperplanes; ++i) {
    pow2[i] = pow(2, i);
  }

  seq = kseq_init(fp);
  hyperplanes = eigg_hyperplane_ary_init(num_hyperplanes, kmer_len);
  encode_nt   = eigg_encode_ary_init();
  tommy_array_init(hash_bucket_names);
  tommy_hashlin_init(hash_bucket_counts);

  /* eigg_print_hyperplanes(hyperplanes, num_hyperplanes, kmer_len); */

  while ((l = kseq_read(seq)) >= 0) {
    s1_name = strdup(seq->name.s);
    s1_qual = strdup(seq->qual.s);
    s1_seq  = strdup(seq->seq.s);
    s1_seq_len = seq->seq.l;

    l = kseq_read(seq);
    if (l < 0) {
      fprintf(stderr,
              "ERROR -- not enough reads in reverse file %s\n",
              argv[3]);
      return 3;
    }

    num_seqs += 2;
    if ((num_seqs % 1000) == 0 && num_seqs != 0) {
      fprintf(stderr, "LOG -- reading seqs: %lu\r", num_seqs);
    }

    if (strlen(s1_seq) < kmer_len) { /* the sequence is too short for the
                                        kmer len */
      fprintf(stderr, "LOG -- Seq: %s, Len: %lu is too short\n",
              s1_name,
              s1_seq_len);

      /* do something to handle it */
      return 4;
    }

    if (seq->seq.l < kmer_len) { /* the sequence is too short for the
                                    kmer len */
      fprintf(stderr, "LOG -- Seq: %s, Len: %lu is too short\n",
              seq->name.s,
              seq->seq.l);

      /* do something to handle it */
      return 4;
    }

    struct Kmers* for_kmers =
      eigg_kmers_new(s1_seq, s1_seq_len, kmer_len);
    double* hashed_for_kmers =
      malloc(for_kmers->num_kmers * sizeof(double));

    /* the rev read might be different lenght */
    struct Kmers* rev_kmers =
      eigg_kmers_new(seq->seq.s, seq->seq.l, kmer_len);
    double* hashed_rev_kmers =
      malloc(rev_kmers->num_kmers * sizeof(double));

    /* hash kmers for forward read */
    for (i = 0; i < for_kmers->num_kmers; ++i) {
      eigg_encode_kmer(encoded_kmer,
                       encode_nt,
                       for_kmers->kmers[i],
                       kmer_len);

      hashed_kmer = eigg_hash_encoded_kmer(hyperplanes,
                                           num_hyperplanes,
                                           encoded_kmer,
                                           kmer_len,
                                           pow2);

      /* have we seen the hashed_kmer in this sample yet? */
      hash_bucket_count = tommy_hashlin_search(hash_bucket_counts,
                                               hash_bucket_count_compare,
                                               &hashed_kmer,
                                               tommy_inthash_u64(hashed_kmer));

      if (hash_bucket_count) { /* found, increment the count */
        /* TODO worry about overflows? */
        ++hash_bucket_count->count;
      } else { /* not found, add it with a count of 1 */
        ++num_seen_hash_buckets;

        /* track the order */
        /* TODO this could be shared with the name obj in the hash instead? */
        unsigned long* hash_bucket = malloc(sizeof(unsigned long));
        assert(hash_bucket != NULL);
        hash_bucket[0] = hashed_kmer;
        tommy_array_insert(hash_bucket_names, hash_bucket);

        struct hash_bucket_count_t* hb_count =
          malloc(sizeof(struct hash_bucket_count_t));
        assert(hb_count != NULL);

        hb_count->hash_bucket = hashed_kmer;
        hb_count->count = 1;

        tommy_hashlin_insert(hash_bucket_counts,
                             &hb_count->node,
                             hb_count,
                             tommy_inthash_u64(hb_count->hash_bucket));
      }

      /* ++hashed_kmer_counts[(unsigned long)hashed_kmer]; */

      hashed_for_kmers[i] = hashed_kmer;
    }

    /* hash kmers for reverse read */
    for (i = 0; i < rev_kmers->num_kmers; ++i) {
      eigg_encode_kmer(encoded_kmer,
                       encode_nt,
                       rev_kmers->kmers[i],
                       kmer_len);

      hashed_kmer = eigg_hash_encoded_kmer(hyperplanes,
                                           num_hyperplanes,
                                           encoded_kmer,
                                           kmer_len,
                                           pow2);

      /* have we seen the hashed_kmer in this sample yet? */
      hash_bucket_count = tommy_hashlin_search(hash_bucket_counts,
                                               hash_bucket_count_compare,
                                               &hashed_kmer,
                                               tommy_inthash_u64(hashed_kmer));

      if (hash_bucket_count) { /* found, increment the count */
        /* TODO worry about overflows? */
        ++hash_bucket_count->count;
      } else { /* not found, add it with a count of 1 */
        ++num_seen_hash_buckets;

        /* track the order */
        /* TODO this could be shared with the name obj in the hash instead? */
        unsigned long* hash_bucket = malloc(sizeof(unsigned long));
        assert(hash_bucket != NULL);
        hash_bucket[0] = hashed_kmer;
        tommy_array_insert(hash_bucket_names, hash_bucket);

        struct hash_bucket_count_t* hb_count =
          malloc(sizeof(struct hash_bucket_count_t));
        assert(hb_count != NULL);

        hb_count->hash_bucket = hashed_kmer;
        hb_count->count = 1;

        tommy_hashlin_insert(hash_bucket_counts,
                             &hb_count->node,
                             hb_count,
                             tommy_inthash_u64(hb_count->hash_bucket));
      }

      /* ++hashed_kmer_counts[(unsigned long)hashed_kmer]; */

      hashed_rev_kmers[i] = hashed_kmer;
    }


    /* each read pair will have a all the kmers of its partner catted
       together with its own */

    /* eigg_print_hashed_record(s1_name, */
    /*                          s1_seq, */
    /*                          s1_qual, */
    /*                          kmer_len, */
    /*                          for_kmers->num_kmers, */
    /*                          rev_kmers->num_kmers, */
    /*                          hashed_for_kmers, */
    /*                          hashed_rev_kmers); */

    /* eigg_print_hashed_record(seq->name.s, */
    /*                          seq->seq.s, */
    /*                          seq->qual.s, */
    /*                          kmer_len, */
    /*                          for_kmers->num_kmers, */
    /*                          rev_kmers->num_kmers, */
    /*                          hashed_for_kmers, */
    /*                          hashed_rev_kmers); */

    /* for (unsigned long i = 0; i < for_kmers->num_kmers; ++i) { */
    /*   fprintf(stdout, */
    /*           "%lu\n", */
    /*           (unsigned long)hashed_for_kmers[i]); */
    /* } */
    /* for (unsigned long i = 0; i < rev_kmers->num_kmers; ++i) { */
    /*   fprintf(stdout, */
    /*           "%lu\n", */
    /*           (unsigned long)hashed_rev_kmers[i]); */
    /* } */

    free(hashed_rev_kmers);
    free(hashed_for_kmers);
    free(s1_name);
    free(s1_seq);
    free(s1_qual);
    eigg_kmers_destroy(rev_kmers);
    eigg_kmers_destroy(for_kmers);
  }
  /* fprintf(stderr, "\n"); */

  /* print kmer counts: outfile is hash_bucket count */
  /* unsigned long count = 0; */

  /* rows, colums, entries */
  fprintf(stdout,
          "1 %lu %lu\n",
          num_hash_buckets,
          num_seen_hash_buckets);
  /* for (unsigned long i = 0; i < num_hash_buckets; ++i) { */
  /*   if ((i % 100000000) == 0 && i != 0 ) { */
  /*     count += 100; */
  /*     fprintf(stderr, */
  /*             "LOG -- processing hash bin counts: %lu million\r", */
  /*             count); */
  /*   } */

  /*   if (hashed_kmer_counts[i] != 0) { */
  /*     fprintf(stdout, */
  /*             "%lu %lu\n", */
  /*             i, */
  /*             hashed_kmer_counts[i]); */
  /*   } */
  /* } */
  /* fprintf(stderr, "\n"); */

  unsigned long* hb_name;
  for (i = 0; i < num_seen_hash_buckets; ++i) {
    if ((i % 100000) == 0 && i != 0) {
      fprintf(stderr,
              "LOG -- processing hash bin counts: %lu\r", i);
    }

    hb_name = (unsigned long*)tommy_array_get(hash_bucket_names, i);
    hash_bucket_count = tommy_hashlin_search(hash_bucket_counts,
                                             hash_bucket_count_compare,
                                             hb_name,
                                             tommy_inthash_u64(hb_name[0]));


    if (hb_name) { /* the hash_bucket is in the hash */
      assert(hb_name[0] == hash_bucket_count->hash_bucket);

      /* sort of like the matrix market format, except the row is not
         printed because we wont know the row until the next step. */
      fprintf(stdout,
              "%lu %lu\n",
              hb_name[0],
              hash_bucket_count->count);
    } else { /* it is missing, but it should be in there */
      fprintf(stderr,
              "Something went wrong! Each bin should be present in bin_counts\n");
      fprintf(stderr,
              "Bin name: %lu\n"
              "Possible hash buckets: %lu\n"
              "Seen hash buckets: %lu\n"
              "Hash bucket num: %lu\n",
              hb_name[0],
              num_hash_buckets,
              num_seen_hash_buckets,
              i);

      return 5;
    }
  }


  kseq_destroy(seq);
  gzclose(fp);
  eigg_hyperplane_ary_destroy(hyperplanes, num_hyperplanes);
  free(encoded_kmer);
  eigg_encode_ary_destroy(encode_nt);
  free(pow2);
  /* free(hashed_kmer_counts); */

  for (tommy_size_t s = 0; s < tommy_array_size(hash_bucket_names); ++s) {
    free(tommy_array_get(hash_bucket_names, s));
  }
  tommy_array_done(hash_bucket_names);
  free(hash_bucket_names);

  tommy_hashlin_foreach(hash_bucket_counts,
                        free);
  tommy_hashlin_done(hash_bucket_counts);
  free(hash_bucket_counts);

  return 0;
}
