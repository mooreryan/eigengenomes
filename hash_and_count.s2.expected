16
2 1
8 3
11 1
12 2
13 1
